cluster-autoscaler/core/resource_slack.go

```diff
@@ -0,0 +1,227 @@
+			"cluster size = %v, cluster cpu sum = %s cores, cluster memory sum = %s, avg milli cpu per node = %v, avg memory per node = %.1fGB, "+
+			"cluster total milli cpu = %v, cluster total memory = %.1fGB, cluster extra milli cpu = %v, cluster extra memory = %.1fGB, "+
+			"max pod milli cpu = %d, max pod memory = %.1fGB, remaining cluster extra milli cpu = %d, remaining cluster extra memory = %.1fGB, "+
+			"min node milli cpu = %d, min node memory = %.1fGB, "+
+			"milli cpu per extra placeholder pod = %d, memory per extra placeholder pod = %.1fGB, extra placeholder pods count = %v",
+		clusterSize,
+		clusterCpuSum.String(),
+		clusterMemorySum.String(),
+		avgNodeMilliCPU,
+		toGB(avgNodeMemory),
+		clusterMilliCPU,
+		toGB(clusterMemory),
+		clusterExtraMilliCPU,
+		toGB(clusterExtraMemory),
+		maxPodRequestedMilliCPU,
+		toGB(float64(maxPodRequestedMemory)),
+		remainingClusterExtraMilliCPU,
+		toGB(remainingClusterExtraMemory),
+		minNodeMilliCPU,
+		toGB(float64(minNodeMemory)),
+		ppMilliCPU,
+		toGB(float64(ppMemory)),
+		extraPlaceholderPodsCount,
+	)
+
+	nodeStickiness := placeholderpod.NodeStickiness{
+		NodeSelector:                 map[string]string{},
+		PodAntiAffinityRequiredTerms: []apiv1.PodAffinityTerm{},
+		NodeAffinityRequiredTerms:    nil,
+	}
+
+	largestCPU := placeholderpod.New(
+		int64(maxPodRequestedMilliCPU),
+		int64(0),
+		nodeStickiness,
+	)
+
+	largestMemory := placeholderpod.New(
+		int64(0),
+		int64(maxPodRequestedMemory),
+		nodeStickiness,
+	)
+
+	standard := placeholderpod.New(
+		int64(ppMilliCPU),
+		int64(ppMemory),
+		nodeStickiness,
+	)
+
+	// TODO Create at least one placeholder pod for every occurrence of a set of nodeSelector + affinity + tolerations
+	// so that even pods demanding such specific requirements on nodes can have extra capacity
+	// TODO Pods dedicated to master nodes should be ignored as master nodes shouldn't be auto-scaled
+	placeholderPods := placeholderpod.CreateReplicaSets(
+		placeholderpod.ReplicaSet{Name: "largestcpu", Count: 1, PodSpec: largestCPU},
+		placeholderpod.ReplicaSet{Name: "largestmemory", Count: 1, PodSpec: largestMemory},
+		placeholderpod.ReplicaSet{Name: "standard", Count: extraPlaceholderPodsCount, PodSpec: standard},
+	)
+
+	glog.V(4).Infof("Created %d placeholder pods", len(placeholderPods))
+
+	return schedulePlaceholderPods(placeholderPods, readyNodes, allScheduled, predicateChecker)
+}
+
+// schedulePlaceholderPods try to fit placeholder pods to actual nodes and virtually assign pods
+// to the fitted nodes so that CA keeps more nodes than actually necessary
+func schedulePlaceholderPods(placeholderPods []*apiv1.Pod, nodes []*apiv1.Node, scheduledPods []*apiv1.Pod, predicateChecker *simulator.PredicateChecker) ([]*apiv1.Pod, []*apiv1.Pod, error) {
+	scheduledPlaceholderPods := []*apiv1.Pod{}
+	unscheduledPlaceholderPods := []*apiv1.Pod{}
+	nodeNameToNodeInfo := createNodeNameToInfoMap(scheduledPods, nodes)
+
+	for _, pod := range placeholderPods {
```

Comment:
On large cluster there is a huge number of pods to be scheduled. And the pods are mostly the same. If a pod from one replica set X failed to schedule on node A then there is no point to in trying to fit the next pod from X on A. Thus there should be just a single loop over all nodes. 

Moreover. the scale down performance strongly depends that the pods don't change every iteration. So if you placed placeholder-X-123 on node A then in the next iteration it should also land on A or the scale down performance will be heavily impacted.